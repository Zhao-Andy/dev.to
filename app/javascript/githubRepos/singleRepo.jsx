import { h, Component } from 'preact';
import PropTypes from 'prop-types';

export class SingleRepo extends Component {
  constructor(props) {
    super(props);
    const { selected } = this.props;
    this.state = { selected };
  }

  forkLabel = () => {
    const { fork } = this.props;
    if (fork) {
      return <span className="github-repo-fork">fork</span>;
    }
    return null;
  };

  submitRepo = () => {
    const { selected } = this.state;
    const { githubIdCode } = this.props;
    this.setState({ selected: !selected });
    const csrfToken = document.querySelector("meta[name='csrf-token']").content;
    const formData = new FormData();
    const formAttributes = {
      github_id_code: githubIdCode,
      featured: !selected,
    };
    formData.append('github_repo', JSON.stringify(formAttributes));

    fetch('/api/github_repos/update_or_create', {
      method: 'POST',
      headers: {
        'X-CSRF-TOKEN': csrfToken,
      },
      body: formData,
      credentials: 'same-origin',
    })
      .then(response => response.json())
      .then(json => {
        this.setState({ selected: json.featured });
      });
  };

  githubRepoClassName = () => {
    const { selected } = this.state;
    if (selected) {
      return 'github-repo-row github-repo-row-selected';
    }
    return 'github-repo-row';
  };

  render() {
    const { selected } = this.state;
    const { name } = this.props;
    return (
      <div className={this.githubRepoClassName()}>
        <div className="github-repo-row-name">
          {name}
          {this.forkLabel()}
          <button className="cta" type="button" onClick={this.submitRepo}>
            {selected ? 'REMOVE' : 'SELECT'}
          </button>
        </div>
      </div>
    );
  }
}

// export const SingleRepo = ({
//   name,
//   githubIdCode,
//   fork,
//   selected,
// }) => {

//   const forkLabel = () => {
//     if(fork) {
//       return(
//         <span className="github-repo-fork">fork</span>
//       )
//     }
//     return null;
//   }

//   const submitRepo = () => {
//     const csrfToken = document.querySelector("meta[name='csrf-token']").content
//     const formData = new FormData()
//     const formAttributes = {
//       'github_id_code': githubIdCode,
//       'featured': !selected
//     }
//     formData.append('github_repo', JSON.stringify(formAttributes))

//     fetch('/api/github_repos/update_or_create', {
//         method: 'POST',
//         headers: {
//           'X-CSRF-TOKEN': csrfToken,
//         },
//         body: formData,
//         credentials: 'same-origin',
//       })
//       .then(response => response.json())
//       .then(json => {
//         console.log(json)
//         debugger
//       })
//   }

//   const githubRepoClassName = () => {
//     if(selected) {
//       return 'github-repo-row github-repo-row-selected'
//     }
//     return 'github-repo-row'
//   }

//   // make form a conditional variable

//   return(
//     <div className={githubRepoClassName()}>
//       <div className="github-repo-row-name">
//         {name}
//         {forkLabel()}
//         <button className="cta" type="button" onClick={submitRepo}>
//           {selected ? 'REMOVE' : 'SELECT'}
//         </button>
//       </div>
//     </div>
//   )
// }

SingleRepo.propTypes = {
  name: PropTypes.string.isRequired,
  githubIdCode: PropTypes.number.isRequired,
  fork: PropTypes.bool.isRequired,
  selected: PropTypes.bool.isRequired,
};
