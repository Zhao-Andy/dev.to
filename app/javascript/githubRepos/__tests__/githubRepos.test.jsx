import { h } from 'preact';
import { shallow } from 'preact-render-spy';
import fetch from 'jest-fetch-mock';
import { GithubRepos } from '../githubRepos';

global.fetch = fetch;

describe('<GithubRepos />', () => {
  const fakeResponse = JSON.stringify([
    {
      github_id_code: 1,
      name: 'dev.to',
      fork: false,
      selected: false,
    },
    {
      github_id_code: 2,
      name: 'a forked repo',
      fork: true,
      selected: true,
    },
  ]);

  // describe('when there are no repos loaded yet', () => {
  //   it('renders properly', () => {
  //     fetch.mockResponse('[{}]')
  //     // const tree = render(<GithubRepos />)
  //     // expect(tree).toMatchSnapshot();
  //     const context = shallow(<GithubRepos />)
  //     expect(context.find('.loading-repos')[0].attributes.className).
  //       toEqual('github-repos loading-repos')
  //     expect(context.state('repos')).toEqual([]);
  //     expect(context.state('erroredOut')).toEqual(false);
  //   })
  // })

  // describe('when there is an error in the response', () => {
  //   it('renders the error message', () => {
  //     fetch.mockResponse('[{}]')
  //     const context = shallow(<GithubRepos />)
  //     console.log(context)
  //     expect(context.find('.github-repos-errored')[0].attributes.className).
  //       toEqual('github-repos github-repos-errored')
  //     expect(context.state('erroredOut')).toEqual(true);
  //     expect(context.state('repos')).toEqual([]);
  //   })
  // })

  describe('when it successfully gets the repos', () => {
    it('renders properly', () => {
      fetch.mockResponse(fakeResponse);
      const context = shallow(<GithubRepos />);

      // context.render()
      // console.log(context)
      // console.log(context.find('.github-repos'))
      expect(context.state()).toMatchSnapshot();
      // expect(context).toEqual(JSON.parse(fakeResponse))
    });
  });
});
