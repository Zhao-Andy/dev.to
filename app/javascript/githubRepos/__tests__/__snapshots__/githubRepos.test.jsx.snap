// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GithubRepos /> when it successfully gets the repos renders properly 1`] = `
Object {
  "erroredOut": false,
  "repos": Array [],
}
`;
